Use Processing-Timestamp based Gap Detection in TransactionalEventProducer

cstub opened the pull request:
To mitigate the risk of event-loss which might happen if multiple application nodes
with unsynchronized clocks emit events or may occur in certain situations with unusually
long CG-times, the event-timestamp based gap detection is discontinued.

Implement a processing-timestamp based gap detection which uses a batched approach
(i.e. all gaps within an event batch will be persisted at once) providing better latency
per gap.

Commit:
Use Processing-Timestamp based Gap Detection in TransactionalEventProducer

To mitigate the risk of event-loss which might happen if multiple application nodes
with unsynchronized clocks emit events or occur in certain situations with unusually
long CG-times the event-timestamp based gap detection is discontinued.

Implement a processing-timestamp based gap detection which uses a batched approach
(i.e. all gaps within an event batch will be persisted at once) providing better latency
per gap.

volkerstampa commented on src/main/scala/com/rbmhtechnology/calliope/GapDetection.scala:
```diff
@@ -29,36 +28,86 @@ import scala.concurrent.duration.FiniteDuration
-  def gapLess[A: Sequenced : Timestamped](fromSeqNr: Long)(implicit sn: Sequenced[A], ts: Timestamped[A]): A => Boolean = {
-    def olderThan(duration: FiniteDuration)(event: A)(implicit referenceTs: Instant): Boolean =
-      referenceTs.toEpochMilli - ts.timestamp(event).toEpochMilli >= duration.toMillis
+trait BatchGapDetection extends GapDetection {
```
I wonder, if we should document somewhere that `gapLess` and the returned function is not thread safe.

cstub replied:
+1

volkerstampa commented on src/test/scala/com/rbmhtechnology/calliope/GapDetectionSpec.scala:
```diff
@@ -0,0 +1,225 @@
+
+      waitForPersistTimeout()
+
+      filterGapLess(4, events(4 to 9)) mustBe events(4 to 7)
```
Wouldn't it be clearer if this were `events(5 to 7)`?

cstub replied:
+1

volkerstampa commented on src/test/scala/com/rbmhtechnology/calliope/GapDetectionSpec.scala:
```diff
@@ -0,0 +1,225 @@
+
+      waitForPersistTimeout()
+
+      filterGapLess(8, events(8 to 10)) mustBe events(8 to 10)
```
`mustBe events(9 to 10)`

cstub replied:
+1

volkerstampa commented on src/test/scala/com/rbmhtechnology/calliope/GapDetectionSpec.scala:
```diff
@@ -0,0 +1,225 @@
+
+      filterGapLess(8, events(8 to 10)) mustBe events(8 to 10)
+
+      filterGapLess(1, events(1 to 5)) mustBe events(1 to 5)
```
Why are these checks part of this test and why isn't it a single check `filterGapLess(1, events(1 to 10)) mustBe events(1 to 10)`

cstub replied:
+1 should be a single check. 
This check is part of this test bc it should illustrate that once gaps have been persisted (in previous iterations) a subsequent iteration is not delayed by the (now persisted) gaps.

volkerstampa commented on src/test/scala/com/rbmhtechnology/calliope/GapDetectionSpec.scala:
```diff
@@ -0,0 +1,225 @@
+
+      waitForPersistTimeout()
+
+      filterGapLess(4, events(4 to 10)) mustBe events(4 to 8)
```
`events(5 to 7)` instead of `4 to 8`?

cstub replied:
+1

volkerstampa commented on src/test/scala/com/rbmhtechnology/calliope/GapDetectionSpec.scala:
```diff
@@ -0,0 +1,225 @@
+
+      filterGapLess(3, events(3 to 12)) mustBe events(3 to 10)
+
+      waitForPersistTimeout()
```
Given we have no rounding error `sleep(timeout / 4)` should be fine, right?

cstub replied:
+1 but to be safe a value of `timeout / 2` ensures that the whole timeout has elapsed.

volkerstampa replied:
But this is still an entire timeout, right? Or am I missing something?

cstub replied:
Ok, now I got it :) I will update it accordingly.

volkerstampa reviewed (commented):
Just some minor issues. I leave it up to you if you want to fix them.

Commit:
- Document that `BatchGapDetection` is not thread-safe and should only be used for a single event-source
- Use a pinned dispatcher for all read-operations of an `EventReader` to accomodate the use of `BatchGapDetection`

Commit:
- Refactor Gap-Detection tests

Commit:
- Incorporate Review Comments

volkerstampa reviewed (approved):
+1
